packages/task/src/browser/task-service.ts

```diff
@@ -402,10 +433,159 @@ export class TaskService implements TaskConfigurationClient {
+                await this.runTasksGraph(dependentTask, tasks, {
+                    customization: { ...taskCustomization, ...{ problemMatcher: resolvedMatchers } }
+                });
+            }
+        } else if (task.dependsOn && task.dependsOn.length > 0 && ((!task.dependsOrder) || (task.dependsOrder && task.dependsOrder === 'parallel'))) {
+            const dependentTasks: { 'task': TaskConfiguration; 'taskCustomization': TaskCustomization; 'resolvedMatchers': ProblemMatcher[] | undefined }[] = [];
+            for (let i = 0; i < task.dependsOn.length; i++) {
+                // It may be a string (a task label) or a JSON object which represents a TaskIdentifier (e.g. {"type":"npm", "script":"script1"})
+                const taskIdentifier = task.dependsOn[i];
+                const dependentTask = this.getDependentTask(taskIdentifier, tasks);
+                const taskCustomization = await this.getTaskCustomization(dependentTask);
+                const resolvedMatchers = await this.resolveProblemMatchers(dependentTask, taskCustomization);
+                dependentTasks.push({ 'task': dependentTask, 'taskCustomization': taskCustomization, 'resolvedMatchers': resolvedMatchers });
+            }
+            const promises = dependentTasks.map(item =>
+                this.runTasksGraph(item.task, tasks, {
+                    customization: { ...item.taskCustomization, ...{ problemMatcher: item.resolvedMatchers } }
+                })
+            );
+            await Promise.all(promises);
+        }
+        const taskInfo = await this.runTask(task, option);
+        if (taskInfo) {
+            if (task.isBackground) {
+                return Promise.race([this.getBackgroundTaskEndMatch(taskInfo.taskId), this.getExitCode(taskInfo.taskId)]).then(async result => {
+                    if (typeof result === 'number' && result !== 0) {
+                        await this.throwTaskError('The task: ' + task.label + ' terminated with exit code ' + result + '.');
+                    }
+                    return taskInfo;
+                });
+            } else {
+                const exitCode = await this.getExitCode(taskInfo.taskId);
+                if (exitCode !== 0) {
+                    await this.throwTaskError('The task: ' + task.label + ' terminated with exit code ' + exitCode + '.');
+                }
+            }
+        }
+        return taskInfo;
+    }
+
+    async throwTaskError(errorMessage: string): Promise<void> {
+        throw new Error(errorMessage);
+    }
+
+    /**
+     * Creates a graph of dependencies tasks from the root task and verify there is no DAG (Directed Acyclic Graph).
+     * In case of detection of a circular dependency, an error is thrown with a message which describes the detected circular reference.
+     */
+    detectDirectedAcyclicGraph(task: TaskConfiguration, tasknode: TaskNode, tasks: TaskConfiguration[]): void {
```

Comment:
[minor] please use camel casing for variables names: `taskNode` (as described in our [coding guidelines](https://github.com/eclipse-theia/theia/wiki/Coding-Guidelines#camelcase-var).

Reply:
done